wgengine/router/router_linux.go

```diff
@@ -54,24 +54,97 @@ type netfilterRunner interface {
 	HasIPV6NAT() bool
 }
 
+// tableDetector abstracts helpers to detect the firewall mode.
+// It is implemented for testing purposes.
+type tableDetector interface {
+	iptDetect() (int, error)
+	nftDetect() (int, error)
+}
+
+type linuxFWDetector struct{}
+
+// iptDetect returns the number of iptables rules in the current namespace.
+func (l *linuxFWDetector) iptDetect() (int, error) {
+	return linuxfw.DetectIptables()
+}
+
+// nftDetect returns the number of nftables rules in the current namespace.
+func (l *linuxFWDetector) nftDetect() (int, error) {
+	return linuxfw.DetectNetfilter()
+}
+
+// chooseFireWallMode returns the firewall mode to use based on the
+// environment and the system's capabilities.
+func chooseFireWallMode(logf logger.Logf, det tableDetector) (linuxfw.FirewallMode, error) {
+	iptAva, nftAva := true, true
+	iptRuleCount, err := det.iptDetect()
+	logf("router: iptables rule count: %d", iptRuleCount)
+	if err != nil {
+		logf("roter: %v", err)
+		iptAva = false
+	}
+	nftRuleCount, err := det.nftDetect()
+	logf("router: nftables rule count: %d", nftRuleCount)
```

Comment:
We should consider logging both the `iptables` and `nftables` rule counts in the same log line, so that if we want to debug a system we can look at a single log line instead of two different ones that might be separated by some logs from other goroutines.